tool/lib/release_note_classes.dart

```diff
@@ -0,0 +1,146 @@
+      _$ReleaseNotesFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ReleaseNotesToJson(this);
+}
+
+@JsonSerializable()
+class Release {
+  Release({
+    required this.version,
+    required this.sections,
+  }) {
+    for (var element in sections) {
+      _sectionMap[element.name] = element;
+    }
+  }
+
+  final SemanticVersion version;
+  List<ReleaseSection> sections = [];
+  final Map<String, ReleaseSection> _sectionMap = {};
+
+  void addNote(String sectionName, ReleaseNote note) {
+    _sectionMap[sectionName]!.notes.add(note);
+  }
+
+  String toMarkdown() {
+    String markdown = '';
+    markdown += '# DevTools $version release notes\n\n';
+    for (var section in sections) {
+      markdown += '# ${section.name}\n\n';
+      for (var note in section.notes) {
+        markdown += note.toMarkdown();
+      }
+      markdown += '\n';
+    }
+    return markdown;
+  }
+
+  factory Release.fromJson(Map<String, dynamic> json) =>
+      _$ReleaseFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ReleaseToJson(this);
+}
+
+@JsonSerializable()
+class ReleaseSection {
+  ReleaseSection({
+    required this.name,
+    List<ReleaseNote>? notes,
+  }) : notes = notes ?? [];
+
+  ReleaseSection.emptyNotes({
+    required this.name,
+  }) : notes = [];
+
+  final String name;
+  final List<ReleaseNote> notes;
+
+  factory ReleaseSection.fromJson(Map<String, dynamic> json) =>
+      _$ReleaseSectionFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ReleaseSectionToJson(this);
+}
+
+@JsonSerializable()
+class ReleaseNote {
+  ReleaseNote({
+    required this.message,
+    this.imageNames,
+    this.githubPullRequestUrls,
+  }) {
+    if (imageNames != null) {
+      for (var name in imageNames!) {
+        final path = "release_notes/files/$name";
+        //TODO: get the proper path?
+        if (!File(path).existsSync()) {
+          throw Exception(
+              "Could not find image file $path for note: \n${toMarkdown()}");
+        }
+      }
+    }
+  }
+
+  List<String>? githubPullRequestUrls;
+  final String message;
+  final List<String>? imageNames;
+
+  String toMarkdown() {
+    String markdown = '';
+    markdown += '- $message';
+    if (githubPullRequestUrls != null) {
+      List<String> prUrls = [];
+      for (var url in githubPullRequestUrls!) {
+        final match = RegExp(
+                r'^https://github.com/flutter/devtools/pull/(?<pr_number>\d+)$')
+            .firstMatch(url);
+        if (match == null) {
+          throw Exception(
+              "Invalid github PR Url($url) found in message: $message}");
+        }
+        final prNumber = match.namedGroup('pr_number');
+        prUrls.add('[#$prNumber]($url)');
+      }
+      markdown += ' - ${prUrls.join(", ")}';
+    }
+
+    markdown += '\n';
+
+    if (imageNames != null) {
+      for (var imageName in imageNames!) {
+        markdown += '![](files/$imageName)\n';
+      }
+    }
+
+    return markdown;
```

Comment:
another idea instead of using markdown += or string interpolation everywhere store a list of `markdownParts` and then call `markdownParts.join()` here